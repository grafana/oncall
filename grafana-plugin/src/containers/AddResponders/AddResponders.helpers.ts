import { ManualAlertGroupFormData } from 'components/ManualAlertGroup/ManualAlertGroup.config';
import { GrafanaTeam } from 'models/grafana_team/grafana_team.types';

import { UserResponders } from './AddResponders.types';

export const prepareForUpdate = (
  selectedUsers: UserResponders,
  selectedTeam?: GrafanaTeam,
  data?: ManualAlertGroupFormData
) => ({
  ...(data || {}),
  team: selectedTeam ? selectedTeam.id : null,
  users: selectedUsers.map(({ important, data: { pk } }) => ({ important, id: pk })),
});
