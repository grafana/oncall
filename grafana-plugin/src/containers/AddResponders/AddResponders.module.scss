.body {
  width: 100%;
  position: relative;
}

.responders-list {
  list-style-type: none;
  margin-bottom: 20px;
  width: 100%;

  & > li .hover-button {
    display: none;
  }

  & > li:hover .hover-button {
    display: inline-flex;
  }

  & > li {
    padding: 10px 12px;
    width: 100%;
  }

  & > li:hover {
    background: var(--background-secondary);
  }
}

.timeline-icon-background {
  width: 28px;
  height: 28px;
  border-radius: 50%;
  background: var(--timeline-icon-background);
  display: flex;
  justify-content: center;
  align-items: center;

  & > img {
    width: 100%;
    height: 100%;
  }

  &--green {
    background: #299c46;
  }
}

.responder-name {
  max-width: 250px;
  overflow: hidden;
  white-space: nowrap;
}

.confirm-participant-invitation-modal {
  max-width: 550px;
}

.confirm-participant-invitation-modal-select {
  display: inline-flex;
  margin: 0 4px;
}

.learn-more-link {
  display: inline-block;
}
