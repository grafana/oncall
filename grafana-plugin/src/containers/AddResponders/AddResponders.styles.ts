import { css } from '@emotion/css';
import { GrafanaTheme2 } from '@grafana/data';

export const getAddRespondersStyles = (theme: GrafanaTheme2) => {
  return {
    content: css`
      width: 100%;
      position: relative;
    `,

    respondersList: css`
      padding-top: 8px;
      list-style-type: none;
      margin-bottom: 20px;
      width: 100%;

      & > li .hover-button {
        display: none;
      }

      & > li:hover .hover-button {
        display: inline-flex;
      }

      & > li {
        padding: 10px 12px;
        width: 100%;
      }

      & > li:hover {
        background: ${theme.colors.background.secondary};
      }
    `,

    alert: css`
      padding-top: 4px;
    `,

    timelineIconBackground: css`
      width: 28px;
      height: 28px;
      border-radius: 50%;
      display: flex;
      justify-content: center;
      align-items: center;
      background: rgba(${theme.isDark ? '70, 76, 84, 1' : '70, 76, 84, 0'});

      & > img {
        width: 100%;
        height: 100%;
      }

      &--green {
        background: #299c46;
      }
    `,

    responderName: css`
      max-width: 250px;
      overflow: hidden;
      white-space: nowrap;
    `,

    confirmParticipantInvitationModal: css`
      max-width: 550px;
    `,

    confirmParticipantInvitationModalSelect: css`
      display: inline-flex;
      margin: 0 4px;
    `,

    learnMoreLink: css`
      display: inline-block;
    `,
  };
};
