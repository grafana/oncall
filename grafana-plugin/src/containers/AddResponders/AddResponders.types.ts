import { SelectableValue } from '@grafana/data';
import { ActionMeta } from '@grafana/ui';

import { ApiSchemas } from 'network/oncall-api/api.types';

export enum NotificationPolicyValue {
  Default = 0,
  Important = 1,
}

export type UserResponder = {
  data: ApiSchemas['UserIsCurrentlyOnCall'];
  important: boolean;
};
export type UserResponders = UserResponder[];

export type ResponderBaseProps = {
  onImportantChange: (value: SelectableValue<number>, actionMeta: ActionMeta) => void | {};
  handleDelete: React.MouseEventHandler<HTMLButtonElement>;
};
