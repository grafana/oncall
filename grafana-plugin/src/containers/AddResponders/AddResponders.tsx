import React, { useState, useCallback, useMemo } from 'react';

import { SelectableValue } from '@grafana/data';
import { HorizontalGroup, Button, Modal, Alert, VerticalGroup, Icon } from '@grafana/ui';
import cn from 'classnames/bind';
import dayjs from 'dayjs';
import { observer } from 'mobx-react';

import { Block } from 'components/GBlock/Block';
import { Text } from 'components/Text/Text';
import { WithPermissionControlTooltip } from 'containers/WithPermissionControl/WithPermissionControlTooltip';
import { UserHelper } from 'models/user/user.helpers';
import { ApiSchemas } from 'network/oncall-api/api.types';
import { useStore } from 'state/useStore';
import { UserActions } from 'utils/authorization/authorization';

import styles from './AddResponders.module.scss';
import { NotificationPolicyValue, UserResponder as UserResponderType } from './AddResponders.types';
import { AddRespondersPopup } from './parts/AddRespondersPopup/AddRespondersPopup';
import { NotificationPoliciesSelect } from './parts/NotificationPoliciesSelect/NotificationPoliciesSelect';
import { TeamResponder } from './parts/TeamResponder/TeamResponder';
import { UserResponder } from './parts/UserResponder/UserResponder';

const cx = cn.bind(styles);

type Props = {
  mode: 'create' | 'update';
  hideAddResponderButton?: boolean;
  existingPagedUsers?: ApiSchemas['AlertGroup']['paged_users'];
  onAddNewParticipant?: (responder: UserResponderType) => Promise<void>;
  generateRemovePreviouslyPagedUserCallback?: (userId: string) => () => Promise<void>;
};

const LearnMoreAboutNotificationPoliciesLink: React.FC = () => (
  <a
    className={cx('learn-more-link')}
    href="https://grafana.com/docs/oncall/latest/notify/#configure-user-notification-policies"
    target="_blank"
    rel="noreferrer"
  >
    <Text type="link">
      <HorizontalGroup spacing="xs">
        Learn more
        <Icon name="external-link-alt" />
      </HorizontalGroup>
    </Text>
  </a>
);

export const AddResponders = observer(
  ({
    mode,
    hideAddResponderButton,
    existingPagedUsers = [],
    onAddNewParticipant,
    generateRemovePreviouslyPagedUserCallback,
  }: Props) => {
    const { directPagingStore } = useStore();
    const { selectedTeamResponder, selectedUserResponders } = directPagingStore;

    const currentMoment = useMemo(() => dayjs(), []);
    const isCreateMode = mode === 'create';

    const [currentlyConsideredUser, setCurrentlyConsideredUser] = useState<ApiSchemas['UserIsCurrentlyOnCall']>(null);
    const [currentlyConsideredUserNotificationPolicy, setCurrentlyConsideredUserNotificationPolicy] =
      useState<NotificationPolicyValue>(NotificationPolicyValue.Default);

    const [popupIsVisible, setPopupIsVisible] = useState(false);
    const [showUserConfirmationModal, setShowUserConfirmationModal] = useState(false);

    const onChangeCurrentlyConsideredUserNotificationPolicy = useCallback(
      ({ value }: SelectableValue<number>) => {
        setCurrentlyConsideredUserNotificationPolicy(value);
      },
      [setCurrentlyConsideredUserNotificationPolicy]
    );

    const closeUserConfirmationModal = useCallback(
      () => setShowUserConfirmationModal(false),
      [setShowUserConfirmationModal]
    );

    const confirmCurrentlyConsideredUser = useCallback(async () => {
      /**
       * if we're in create mode (ie. manually creating an alert group),
       * we need to add the user to the array of selected users
       * otherwise, as soon as the modal is confirmed, we add the user to the pre-existing list of "paged users"
       * for the alert group
       */
      if (isCreateMode) {
        directPagingStore.addUserToSelectedUsers(currentlyConsideredUser);
      } else {
        await onAddNewParticipant({
          important: Boolean(currentlyConsideredUserNotificationPolicy),
          data: currentlyConsideredUser,
        });
      }

      closeUserConfirmationModal();
    }, [
      isCreateMode,
      directPagingStore,
      currentlyConsideredUser,
      currentlyConsideredUserNotificationPolicy,
      closeUserConfirmationModal,
    ]);

    return (
      <>
        <div className={cx('body')}>
          <Block bordered>
            <HorizontalGroup justify="space-between">
              <Text.Title type="primary" level={4}>
                Participants
              </Text.Title>
              {!hideAddResponderButton && (
                <WithPermissionControlTooltip userAction={UserActions.AlertGroupsDirectPaging}>
                  <Button
                    variant="secondary"
                    icon="plus"
                    onClick={() => {
                      setPopupIsVisible(true);
                    }}
                  >
                    {isCreateMode ? 'Invite' : 'Add'}
                  </Button>
                </WithPermissionControlTooltip>
              )}
            </HorizontalGroup>
            {(selectedTeamResponder || existingPagedUsers.length > 0 || selectedUserResponders.length > 0) && (
              <>
                <ul className={cx('responders-list')}>
                  {selectedTeamResponder && (
                    <TeamResponder team={selectedTeamResponder} handleDelete={directPagingStore.resetSelectedTeam} />
                  )}
                  {existingPagedUsers.map((user) => (
                    <UserResponder
                      key={user.pk}
                      onImportantChange={() => {}}
                      disableNotificationPolicySelect
                      handleDelete={generateRemovePreviouslyPagedUserCallback(user.pk)}
                      important={user.important}
                      data={user as unknown as ApiSchemas['UserIsCurrentlyOnCall']}
                    />
                  ))}
                  {selectedUserResponders.map((responder, index) => (
                    <UserResponder
                      key={responder.data.pk}
                      onImportantChange={({ value: important }) =>
                        directPagingStore.updateSelectedUserImportantStatus(index, Boolean(important))
                      }
                      handleDelete={() => directPagingStore.removeSelectedUser(index)}
                      {...responder}
                    />
                  ))}
                  {selectedUserResponders.length > 0 && (
                    <Alert
                      severity="info"
                      title={
                        (
                          <Text type="primary">
                            <LearnMoreAboutNotificationPoliciesLink /> about Default vs Important user personal
                            notification settings
                          </Text>
                        ) as any
                      }
                    />
                  )}
                </ul>
              </>
            )}
          </Block>
          <AddRespondersPopup
            mode={mode}
            visible={popupIsVisible}
            setVisible={setPopupIsVisible}
            existingPagedUsers={existingPagedUsers}
            setCurrentlyConsideredUser={setCurrentlyConsideredUser}
            setShowUserConfirmationModal={setShowUserConfirmationModal}
          />
        </div>
        {showUserConfirmationModal && (
          <Modal
            isOpen
            title="Confirm Participant Invitation"
            onDismiss={closeUserConfirmationModal}
            className={cx('confirm-participant-invitation-modal')}
          >
            <VerticalGroup spacing="md">
              {!isCreateMode && (
                <div>
                  <Text>
                    <Text strong>{currentlyConsideredUser.name || currentlyConsideredUser.username}</Text> (local time{' '}
                    {currentMoment.tz(UserHelper.getTimezone(currentlyConsideredUser)).format('HH:mm')}) will be
                    notified using
                  </Text>
                  <div className={cx('confirm-participant-invitation-modal-select')}>
                    <NotificationPoliciesSelect
                      important={Boolean(currentlyConsideredUserNotificationPolicy)}
                      onChange={onChangeCurrentlyConsideredUserNotificationPolicy}
                    />
                  </div>
                  <Text>notification settings. </Text>
                  <LearnMoreAboutNotificationPoliciesLink />
                </div>
              )}
              {!currentlyConsideredUser.is_currently_oncall && (
                <Alert
                  severity="warning"
                  title="This user is not currently on-call. We don't recommend to page users outside on-call hours."
                />
              )}
              <HorizontalGroup justify="flex-end">
                <Button variant="secondary" onClick={closeUserConfirmationModal}>
                  Cancel
                </Button>
                <Button variant="primary" onClick={confirmCurrentlyConsideredUser} data-testid="confirm-non-oncall">
                  Confirm
                </Button>
              </HorizontalGroup>
            </VerticalGroup>
          </Modal>
        )}
      </>
    );
  }
);
