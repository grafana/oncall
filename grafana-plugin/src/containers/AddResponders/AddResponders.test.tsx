import React from 'react';

import { render } from '@testing-library/react';
import { Provider } from 'mobx-react';

import { AddResponders } from './AddResponders';

jest.mock('./parts/AddRespondersPopup/AddRespondersPopup', () => ({
  __esModule: true,
  AddRespondersPopup: () => <div>AddRespondersPopup</div>,
}));

jest.mock('containers/WithPermissionControl/WithPermissionControlTooltip', () => ({
  WithPermissionControlTooltip: ({ children }) => <div>{children}</div>,
}));

describe('AddResponders', () => {
  const generateRemovePreviouslyPagedUserCallback = jest.fn();

  test.each<'create' | 'update'>(['create', 'update'])('should render properly in %s mode', (mode) => {
    const mockStoreValue = {
      directPagingStore: {
        selectedTeamResponder: null,
        selectedUserResponders: [],
      },
    };

    const component = render(
      <Provider store={mockStoreValue}>
        <AddResponders
          mode={mode}
          generateRemovePreviouslyPagedUserCallback={generateRemovePreviouslyPagedUserCallback}
        />
      </Provider>
    );
    expect(component.container).toMatchSnapshot();
  });

  test.each([true, false])(
    'should properly display the add responders button when hideAddResponderButton is %s',
    (hideAddResponderButton) => {
      const mockStoreValue = {
        directPagingStore: {
          selectedTeamResponder: null,
          selectedUserResponders: [],
        },
      };

      const component = render(
        <Provider store={mockStoreValue}>
          <AddResponders
            mode="create"
            hideAddResponderButton={hideAddResponderButton}
            generateRemovePreviouslyPagedUserCallback={generateRemovePreviouslyPagedUserCallback}
          />
        </Provider>
      );
      expect(component.container).toMatchSnapshot();
    }
  );

  test('should render selected team and users properly', () => {
    const mockStoreValue = {
      directPagingStore: {
        selectedTeamResponder: {
          id: 'asdfasdf',
          avatar_url: 'https://example.com',
          name: 'my test team',
        },
        selectedUserResponders: [
          {
            data: {
              pk: 'mcvnm',
              avatar: 'https://example.com/user123.png',
              username: 'my test user',
            },
          },
          {
            data: {
              pk: 'iuo',
              avatar: 'https://example.com/user456.png',
              username: 'my test user2',
            },
          },
        ],
      },
    };

    const component = render(
      <Provider store={mockStoreValue}>
        <AddResponders
          mode="create"
          existingPagedUsers={[
            {
              pk: 'asdfasdf',
              avatar: 'https://example.com/user9995.png',
              username: 'my test user3',
            } as any,
          ]}
          generateRemovePreviouslyPagedUserCallback={generateRemovePreviouslyPagedUserCallback}
        />
      </Provider>
    );
    expect(component.container).toMatchSnapshot();
  });
});
