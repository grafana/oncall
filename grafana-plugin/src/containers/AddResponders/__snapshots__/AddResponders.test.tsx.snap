// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddResponders should properly display the add responders button when hideAddResponderButton is false 1`] = `
<div>
  <div
    class="body"
  >
    <div
      class="root root_bordered"
    >
      <div
        class="css-1mhys9y-horizontal-group"
        style="width: 100%; height: 100%;"
      >
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <h4
            class="title"
          >
            <span
              class="root text text--primary text--medium"
            >
              Participants
            </span>
          </h4>
        </div>
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <div>
            <button
              class="css-8b29hm-button"
              type="button"
            >
              <div
                class="css-1j2891d-Icon"
              />
              <span
                class="css-1riaxdn"
              >
                Invite
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div>
      AddRespondersPopup
    </div>
  </div>
</div>
`;

exports[`AddResponders should properly display the add responders button when hideAddResponderButton is true 1`] = `
<div>
  <div
    class="body"
  >
    <div
      class="root root_bordered"
    >
      <div
        class="css-1mhys9y-horizontal-group"
        style="width: 100%; height: 100%;"
      >
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <h4
            class="title"
          >
            <span
              class="root text text--primary text--medium"
            >
              Participants
            </span>
          </h4>
        </div>
      </div>
    </div>
    <div>
      AddRespondersPopup
    </div>
  </div>
</div>
`;

exports[`AddResponders should render properly in create mode 1`] = `
<div>
  <div
    class="body"
  >
    <div
      class="root root_bordered"
    >
      <div
        class="css-1mhys9y-horizontal-group"
        style="width: 100%; height: 100%;"
      >
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <h4
            class="title"
          >
            <span
              class="root text text--primary text--medium"
            >
              Participants
            </span>
          </h4>
        </div>
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <div>
            <button
              class="css-8b29hm-button"
              type="button"
            >
              <div
                class="css-1j2891d-Icon"
              />
              <span
                class="css-1riaxdn"
              >
                Invite
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div>
      AddRespondersPopup
    </div>
  </div>
</div>
`;

exports[`AddResponders should render properly in update mode 1`] = `
<div>
  <div
    class="body"
  >
    <div
      class="root root_bordered"
    >
      <div
        class="css-1mhys9y-horizontal-group"
        style="width: 100%; height: 100%;"
      >
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <h4
            class="title"
          >
            <span
              class="root text text--primary text--medium"
            >
              Participants
            </span>
          </h4>
        </div>
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <div>
            <button
              class="css-8b29hm-button"
              type="button"
            >
              <div
                class="css-1j2891d-Icon"
              />
              <span
                class="css-1riaxdn"
              >
                Add
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div>
      AddRespondersPopup
    </div>
  </div>
</div>
`;

exports[`AddResponders should render selected team and users properly 1`] = `
<div>
  <div
    class="body"
  >
    <div
      class="root root_bordered"
    >
      <div
        class="css-1mhys9y-horizontal-group"
        style="width: 100%; height: 100%;"
      >
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <h4
            class="title"
          >
            <span
              class="root text text--primary text--medium"
            >
              Participants
            </span>
          </h4>
        </div>
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <div>
            <button
              class="css-8b29hm-button"
              type="button"
            >
              <div
                class="css-1j2891d-Icon"
              />
              <span
                class="css-1riaxdn"
              >
                Invite
              </span>
            </button>
          </div>
        </div>
      </div>
      <ul
        class="responders-list"
      >
        <li>
          <div
            class="css-1mhys9y-horizontal-group"
            style="width: 100%; height: 100%;"
          >
            <div
              class="css-18qv8yz-layoutChildrenWrapper"
            >
              <div
                class="css-ffyaiw-horizontal-group"
                style="width: 100%; height: 100%;"
              >
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <div
                    class="timeline-icon-background"
                  >
                    <img
                      class="root avatarSize-medium"
                      data-testid="test__avatar"
                      src="https://example.com"
                    />
                  </div>
                </div>
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <span
                    class="root text text--undefined text--medium responder-name"
                  >
                    my test team
                  </span>
                </div>
              </div>
            </div>
            <div
              class="css-18qv8yz-layoutChildrenWrapper"
            >
              <button
                aria-label="Remove responder"
                class="css-17584xm"
                data-testid="team-responder-delete-icon"
                tabindex="0"
                type="button"
              >
                <div
                  class="css-1j2891d-Icon"
                />
              </button>
            </div>
          </div>
        </li>
        <li>
          <div
            class="css-1mhys9y-horizontal-group"
            style="width: 100%; height: 100%;"
          >
            <div
              class="css-18qv8yz-layoutChildrenWrapper"
            >
              <div
                class="css-ffyaiw-horizontal-group"
                style="width: 100%; height: 100%;"
              >
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <div
                    class="timeline-icon-background timeline-icon-background--green"
                  >
                    <img
                      class="root avatarSize-medium"
                      data-testid="test__avatar"
                      src="https://example.com/user9995.png"
                    />
                  </div>
                </div>
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <span
                    class="root text text--undefined text--medium responder-name"
                  >
                    my test user3
                  </span>
                </div>
              </div>
            </div>
            <div
              class="css-18qv8yz-layoutChildrenWrapper"
            >
              <div
                class="css-ffyaiw-horizontal-group"
                style="width: 100%; height: 100%;"
              >
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <div
                    class="css-e49k3t-input-wrapper select css-8k5qe3-SelectContainer"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-2-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="css-1i88p6p"
                    >
                      <div
                        class="css-1q0c0d5-grafana-select-value-container"
                      >
                        <div
                          class=" css-1n8tjau-placeholder"
                          id="react-select-2-placeholder"
                        >
                          Select...
                        </div>
                        <input
                          aria-autocomplete="list"
                          aria-describedby="react-select-2-placeholder"
                          aria-expanded="false"
                          aria-haspopup="true"
                          aria-readonly="true"
                          class="css-mohuvp-dummyInput-DummyInput"
                          disabled=""
                          id="react-select-2-input"
                          inputmode="none"
                          role="combobox"
                          tabindex="0"
                          value=""
                        />
                      </div>
                      <div
                        class="css-zyjsuv-input-suffix"
                      >
                        <div
                          class="css-1j2891d-Icon"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <button
                    aria-label="Remove responder"
                    class="css-17584xm"
                    data-testid="user-responder-delete-icon"
                    tabindex="0"
                    type="button"
                  >
                    <div
                      class="css-1j2891d-Icon"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </li>
        <li>
          <div
            class="css-1mhys9y-horizontal-group"
            style="width: 100%; height: 100%;"
          >
            <div
              class="css-18qv8yz-layoutChildrenWrapper"
            >
              <div
                class="css-ffyaiw-horizontal-group"
                style="width: 100%; height: 100%;"
              >
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <div
                    class="timeline-icon-background timeline-icon-background--green"
                  >
                    <img
                      class="root avatarSize-medium"
                      data-testid="test__avatar"
                      src="https://example.com/user123.png"
                    />
                  </div>
                </div>
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <span
                    class="root text text--undefined text--medium responder-name"
                  >
                    my test user
                  </span>
                </div>
              </div>
            </div>
            <div
              class="css-18qv8yz-layoutChildrenWrapper"
            >
              <div
                class="css-ffyaiw-horizontal-group"
                style="width: 100%; height: 100%;"
              >
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <div
                    class="css-1nmqu8c-input-wrapper select css-8k5qe3-SelectContainer"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-3-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="css-1i88p6p"
                    >
                      <div
                        class="css-1q0c0d5-grafana-select-value-container"
                      >
                        <div
                          class=" css-1n8tjau-placeholder"
                          id="react-select-3-placeholder"
                        >
                          Select...
                        </div>
                        <input
                          aria-autocomplete="list"
                          aria-describedby="react-select-3-placeholder"
                          aria-expanded="false"
                          aria-haspopup="true"
                          aria-readonly="true"
                          class="css-mohuvp-dummyInput-DummyInput"
                          id="react-select-3-input"
                          inputmode="none"
                          role="combobox"
                          tabindex="0"
                          value=""
                        />
                      </div>
                      <div
                        class="css-zyjsuv-input-suffix"
                      >
                        <div
                          class="css-1j2891d-Icon"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <button
                    aria-label="Remove responder"
                    class="css-17584xm"
                    data-testid="user-responder-delete-icon"
                    tabindex="0"
                    type="button"
                  >
                    <div
                      class="css-1j2891d-Icon"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </li>
        <li>
          <div
            class="css-1mhys9y-horizontal-group"
            style="width: 100%; height: 100%;"
          >
            <div
              class="css-18qv8yz-layoutChildrenWrapper"
            >
              <div
                class="css-ffyaiw-horizontal-group"
                style="width: 100%; height: 100%;"
              >
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <div
                    class="timeline-icon-background timeline-icon-background--green"
                  >
                    <img
                      class="root avatarSize-medium"
                      data-testid="test__avatar"
                      src="https://example.com/user456.png"
                    />
                  </div>
                </div>
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <span
                    class="root text text--undefined text--medium responder-name"
                  >
                    my test user2
                  </span>
                </div>
              </div>
            </div>
            <div
              class="css-18qv8yz-layoutChildrenWrapper"
            >
              <div
                class="css-ffyaiw-horizontal-group"
                style="width: 100%; height: 100%;"
              >
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <div
                    class="css-1nmqu8c-input-wrapper select css-8k5qe3-SelectContainer"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-4-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="css-1i88p6p"
                    >
                      <div
                        class="css-1q0c0d5-grafana-select-value-container"
                      >
                        <div
                          class=" css-1n8tjau-placeholder"
                          id="react-select-4-placeholder"
                        >
                          Select...
                        </div>
                        <input
                          aria-autocomplete="list"
                          aria-describedby="react-select-4-placeholder"
                          aria-expanded="false"
                          aria-haspopup="true"
                          aria-readonly="true"
                          class="css-mohuvp-dummyInput-DummyInput"
                          id="react-select-4-input"
                          inputmode="none"
                          role="combobox"
                          tabindex="0"
                          value=""
                        />
                      </div>
                      <div
                        class="css-zyjsuv-input-suffix"
                      >
                        <div
                          class="css-1j2891d-Icon"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="css-18qv8yz-layoutChildrenWrapper"
                >
                  <button
                    aria-label="Remove responder"
                    class="css-17584xm"
                    data-testid="user-responder-delete-icon"
                    tabindex="0"
                    type="button"
                  >
                    <div
                      class="css-1j2891d-Icon"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </li>
        <div
          aria-label="[object Object]"
          class="css-10yjoiw"
          data-testid="data-testid Alert info"
          role="status"
        >
          <div
            class="css-1td7znu"
          >
            <div
              class="css-ufgc62"
            >
              <div
                class="css-tluiue"
              >
                <div
                  class="css-1j2891d-Icon"
                />
              </div>
            </div>
            <div
              class="css-1gmwkrf"
            >
              <span
                class="css-9om60p"
              >
                <span
                  class="root text text--primary text--medium"
                >
                  <a
                    class="learn-more-link"
                    href="https://grafana.com/docs/oncall/latest/notify/#configure-user-notification-policies"
                    rel="noreferrer"
                    target="_blank"
                  >
                    <span
                      class="root text text--link text--medium"
                    >
                      <div
                        class="css-ffyaiw-horizontal-group"
                        style="width: 100%; height: 100%;"
                      >
                        <div
                          class="css-12kn7ff-layoutChildrenWrapper"
                        >
                          Learn more
                        </div>
                        <div
                          class="css-12kn7ff-layoutChildrenWrapper"
                        >
                          <div
                            class="css-1j2891d-Icon"
                          />
                        </div>
                      </div>
                    </span>
                  </a>
                   about Default vs Important user personal notification settings
                </span>
              </span>
            </div>
          </div>
        </div>
      </ul>
    </div>
    <div>
      AddRespondersPopup
    </div>
  </div>
</div>
`;
